src/Microsoft.AspNetCore.Razor/CodeGenerators/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs

```diff
@@ -86,10 +91,87 @@ protected override void Visit(TagHelperChunk chunk)
                 }
             }
 
+            if (!Context.Host.DesignTimeMode)
+            {
+                PreAllocateUnboundTagHelperAttributes(chunk);
+            }
+
             // We need to dive deeper to ensure we pick up any nested tag helpers.
             Accept(chunk.Children);
         }
 
+        private void PreAllocateUnboundTagHelperAttributes(TagHelperChunk chunk)
+        {
+            var boundAttributes = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            for (var i = 0; i < chunk.Attributes.Count; i++)
+            {
+                var attribute = chunk.Attributes[i];
+                var associatedDescriptors = chunk.Descriptors.Where(descriptor =>
```

Comment:
Doesn't look like this actually needs the descriptors. You could do:
`var hasAssociatedDescriptors = chunk.Descriptors.Any(..);` and use that in the if later on.